src/nvim/message.c

```diff
@@ -3022,6 +3022,13 @@ void msg_clr_eos_force(void)
     msg_row = msg_grid_pos;
   }
 
+  if (update_cmdline && msg_col > 0) {
+    for (int i = 0; i < msg_col; i++) {
+      grid_fill(&msg_grid_adj, msg_row, msg_row + 1, i,  i + 1, (int)msg_grid.chars[i],
```

Comment:
too much of a hack, better to just let existing cells be what they are.

Reply:
this will make exist cell use new attr . otherwise they are not take effect

Reply:
hmm this will make terminal/buffer_spec failed

Reply:
this more hack..can make test pass..but I think this is not allowed.

```diff
diff --git a/src/nvim/message.c b/src/nvim/message.c
index bb5bd9753..852223157 100644
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -3022,7 +3022,7 @@ void msg_clr_eos_force(void)
     msg_row = msg_grid_pos;
   }

-  if (need_clr_eos && msg_col > 0) {
+  if (need_clr_eos && msg_col > 0 && msg_grid.chars[0] != '-') {
     for (int i = 0; i < msg_col; i++) {
       grid_fill(&msg_grid_adj, msg_row, msg_row + 1, i,  i + 1, (int)msg_grid.chars[i],
                 (int)msg_grid.chars[i], HL_ATTR(HLF_MSG));
```

Reply:
It is too late to change highlights here, also for the reasons nwounkn mentioned. I rather think that this is missing:
```
diff --git a/src/nvim/message.c b/src/nvim/message.c
index c1acef8b0d..d73f9562e4 100644
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -32,6 +32,7 @@
 #include "nvim/globals.h"
 #include "nvim/grid.h"
 #include "nvim/highlight.h"
+#include "nvim/highlight_group.h"
 #include "nvim/indent.h"
 #include "nvim/input.h"
 #include "nvim/keycodes.h"
@@ -1418,6 +1419,10 @@ void msg_start(void)
     need_fileinfo = false;
   }
 
+  if (need_highlight_changed) {
+    highlight_changed();
+  }
+
   if (need_clr_eos || (p_ch == 0 && redrawing_cmdline)) {
     // Halfway an ":echo" command and getting an (error) message: clear
     // any text from the command.
```

Reply:
oh yep. in msg_start refresh highlight will be a good point :) get it.